File: javaCode/SmithGame/src/me/haved/engine/core/GameWindow.java
package me.haved.engine.core;

import org.lwjgl.glfw.*;
import org.lwjgl.opengl.*;
 
import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.system.MemoryUtil.*;

import org.lwjgl.Version;

public class GameWindow {
	private GLFWErrorCallback 		errorCallback;
	private GLFWKeyCallback 		keyCallback;
	private GLFWWindowSizeCallback 	sizeCallback;
	
	private long windowHandle;
	
	private Game game;
	
	private String title;
	private int width, height, samples;
	private boolean vsync;
	
	public GameWindow(Game game, String title, int width, int height, int samples, boolean vsync) {
		this.game       = game;
		this.title 		= title;
		this.width 		= width;
		this.height 	= height;
		this.samples 	= samples;
		this.vsync 		= vsync;
	}
	
	public void Run(int frameRate) {
		System.out.println("LWJGL Version " + Version.getVersion());
		try {
			glfwSetErrorCallback(errorCallback = GLFWErrorCallback.createPrint(System.err));
			if(glfwInit()!=GLFW_TRUE)
				throw new IllegalStateException("Unable to initialize GLFW");
			
			glfwDefaultWindowHints();
			glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
			glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE);
			glfwWindowHint(GLFW_SAMPLES, samples);
			glfwWindowHint(GLFW_REFRESH_RATE, GLFW_DONT_CARE);
			
			windowHandle = glfwCreateWindow(width, height, title, NULL, NULL);
			if(windowHandle==NULL)
				throw new RuntimeException("Failed to create GLFW window");
			
			glfwSetKeyCallback(windowHandle, keyCallback = new GLFWKeyCallback() {
				@Override
				public void invoke(long window, int key, int scancode, int action, int mods) {
					game.OnKeyEvent(key, scancode, action, mods);
				}
			});
			
			glfwSetWindowSizeCallback(windowHandle, sizeCallback = new GLFWWindowSizeCallback() {
				@Override
				public void invoke(long window, int width, int height) {
					GameWindow.this.width = width;
					GameWindow.this.height = height;
					GameWindow.this.game.Resize(width, height);
				}
			});
			
			GLFWVidMode vidmode = glfwGetVideoMode(glfwGetPrimaryMonitor());
			
			glfwSetWindowPos(windowHandle, (vidmode.width()-width)/2, (vidmode.height()-height)/2);
			
			glfwMakeContextCurrent(windowHandle);
			glfwSwapInterval(vsync?1:0);
			glfwShowWindow(windowHandle);
			
			GL.createCapabilities();
			
			game.Init(this);
			game.Resize(width, height);
			
			final float msPerFrame = frameRate==0?16:(1000f/frameRate);
			int framesAgo=0;
			long startTime=System.currentTimeMillis();
			while(glfwWindowShouldClose(windowHandle)==GLFW_FALSE) {
				Time.updateDeltaTime();
				game.Update();
				game.Render();
				glfwPollEvents();
				
				if(frameRate > 0) {
					framesAgo++;
					long passedTime = System.currentTimeMillis() - startTime;
					long wantedTime = (long)(framesAgo*msPerFrame);
					long extraTime = wantedTime-passedTime;
					if(extraTime>0)
						try {
							Thread.sleep(extraTime);
							Thread.yield();
						}
						catch(Exception e) {
							e.printStackTrace();
						}
					if(framesAgo>=frameRate) {
						startTime+=1000;
						framesAgo=0;
					}
				}
			}
			
			glfwDestroyWindow(windowHandle);
			keyCallback.release();
			sizeCallback.release();
		} finally {
			glfwTerminate();
			errorCallback.release();
		}
	}
	
	public void MakeCurrentContext() {
		glfwMakeContextCurrent(windowHandle);
	}
	
	public void SwapBuffers() {
		glfwSwapBuffers(windowHandle);
	}
	
	public void SetWindowClose() {
		glfwSetWindowShouldClose(windowHandle, GLFW_TRUE);
	}
}

File: cppCode/SmithGame/src/GameWindow.cpp
#include "GameWindow.h"
#include <iostream>
#include "Time.h"
#include <glm/glm.hpp>
#include <SDL2/SDL.h>

GameWindow* GameWindow::instance=NULL;

GameWindow::GameWindow(Game* game, const char* title, int width, int height, int samples, bool vsync)
{
    GameWindow::instance=this;
	m_game = game;
	m_title = title;
	m_width = width;
	m_height = height;
	m_samples = samples;
	m_vsync = vsync;
}

int loadThread(void* arg) {
    GameWindow* gameWindow = GameWindow::instance;
    GLFWwindow* threadContext = gameWindow->GetThreadContext();
    glfwMakeContextCurrent(threadContext);
    Game* game = gameWindow->GetGame();

    while(!glfwWindowShouldClose(threadContext)) {
        game->LoaderUpdate();
        SDL_Delay(10);
    }
    std::cout << "Using NO context!" << std::endl;
    glfwMakeContextCurrent(NULL);
    return 0;
}

void error_callback(int error, const char* description) {
    puts(description);
}

void window_resize_callback(GLFWwindow* window, int width, int height) {
    GameWindow::instance->GetGame()->OnResize(width, height);
}

void window_keyevent_callback(GLFWwindow* window, int key, int scancode, int action, int mods) {
    GameWindow::instance->GetGame()->OnKeyEvent(key, action);
}

int GameWindow::Run(uint32_t frameRate) {
	std::cout << "Running application!" << std::endl;

    //====================================================================INIT of SDL and GLFW
	if (SDL_Init(SDL_INIT_TIMER) < 0) {
		std::cerr << "SDL failed to Init! Aborting!" << std::endl;
		return -1;
	}
	glfwSetErrorCallback(error_callback);
    if(!glfwInit()) {
        std::cerr << "GLFW failed to init! Aborting!" << std::endl;
        exit(EXIT_FAILURE);
    }
    //====================================================================GLFW render window window creation
	glfwDefaultWindowHints();
    glfwWindowHint(GLFW_VISIBLE, GL_FALSE);
    glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);
    glfwWindowHint(GLFW_SAMPLES, m_samples);
    m_renderWindow = glfwCreateWindow(m_width, m_height, m_title, NULL, NULL);
    if (!m_renderWindow) {
		std::cerr << "GLFW window creation failed!" << std::endl;
		glfwTerminate();
		exit(EXIT_FAILURE);
	}
    const GLFWvidmode* vidmode = glfwGetVideoMode(glfwGetPrimaryMonitor()); //Center the renderWindow on the screen
    glfwSetWindowPos(m_renderWindow, (vidmode->width-m_width)/2, (vidmode->height-m_height)/2);
    //=====================================Render window callbacks
    glfwSetWindowSizeCallback(m_renderWindow, window_resize_callback);
    glfwSetKeyCallback(m_renderWindow, window_keyevent_callback);
    //=====================================Thread window
    m_threadWindow = glfwCreateWindow(80, 60, "ThreadWindow", NULL, m_renderWindow);
    if (!m_threadWindow) {
		std::cerr << "GLFW thread window creation failed!" << std::endl;
		glfwDestroyWindow(m_renderWindow);
		glfwTerminate();
		exit(EXIT_FAILURE);
	}
    //====================================================================INIT of GLEW
    glfwMakeContextCurrent(m_renderWindow);

    glewExperimental=true;
    GLenum err = glewInit();
	if(err!=GLEW_OK) {
		std::cerr << "Failed to initialize GLEW!" << glewGetErrorString(err) << std::endl;
		glfwDestroyWindow(m_threadWindow);
		glfwDestroyWindow(m_renderWindow);
		glfwTerminate();
		exit(EXIT_FAILURE);
	}

    //====================================================================GAME INIT and LOOP
	m_game->Init();
	m_game->OnResize(m_width, m_height);
    SDL_Thread* loaderThread = SDL_CreateThread(loadThread, "SmithGame_Loader_thread", NULL);//Making thread
    glfwShowWindow(m_renderWindow);
    uint32_t startMs = SDL_GetTicks();
    uint32_t framesAgo = 0;
    float msPerFrame = 1000.f/frameRate;
	while (!glfwWindowShouldClose(m_renderWindow)) {
        UpdateDeltaTime();
		m_game->NextFrame();
		glfwSwapBuffers(m_renderWindow);
		glfwPollEvents();
        framesAgo++;
        uint32_t wantedTime = startMs+((uint32_t)(framesAgo*msPerFrame));
        uint32_t currentTime = SDL_GetTicks();
        if(wantedTime>currentTime)
            SDL_Delay(wantedTime-currentTime);
        if(framesAgo>=frameRate) {
            framesAgo=0;
            startMs+=1000;
        }
	}
	//====================================================================TELL and wait for the loading thread to stop,
    glfwSetWindowShouldClose(m_threadWindow, GL_TRUE);
    int threadReturnValue;
    SDL_WaitThread(loaderThread, &threadReturnValue);
    //====================================================================DESTROY everytning
    m_game->Destroy();
	glfwDestroyWindow(m_threadWindow);
	glfwDestroyWindow(m_renderWindow);
    glfwTerminate();
	SDL_Quit();
    std::cout << "Program end" << std::endl;
	return 0;
}
